src/main/java/org/elasticsearch/common/xcontent/XContentHelper.java

```diff
@@ -260,11 +260,11 @@ public static boolean update(Map<String, Object> source, Map<String, Object> cha
             if (modified) {
                 continue;
             }
-            if (!checkUpdatesAreUnequal || old == null) {
+            if (!checkUpdatesAreUnequal) {
                 modified = true;
                 continue;
             }
-            modified = !old.equals(changesEntry.getValue());
+            modified = old == null ? changesEntry.getValue() != null : !old.equals(changesEntry.getValue());
```

Comment:
maybe this could be simplified to `modified = !java.util.Objects.equal(old, changesEntry.getValue())`?